src/browser/api/service.ts

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2017 OpenFin Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { applyPendingServiceConnections } from '../api_protocol/api_handlers/service_middleware';
+import { Identity, ServiceIdentity } from '../../shapes';
+import ofEvents from '../of_events';
+import route from '../../common/route';
+
+
+const serviceMap: Map<string, ServiceIdentity> = new Map();
+
+export module Service {
+    export function addEventListener(targetIdentity: Identity, type: string, listener: Function) {
+        // treating like an application so it will work for external connections as well
+        const eventString = route.service(type, targetIdentity.uuid);
+        const errRegex = /^Attempting to call a function in a renderer frame that has been closed or released/;
+        let unsubscribe;
+        let browserWinIsDead;
+
+        const safeListener = (...args: any[]) => {
+            try {
+                listener.call(null, ...args);
+            } catch (err) {
+                browserWinIsDead = errRegex.test(err.message);
```

Comment:
There are a few places where we would have to make changes to make this work for external applications.  I added a comment here to note that this would need to be amended for ext apps, and I made a JIRA ticket for that purpose (https://appoji.jira.com/browse/RUN-3898)